rust/main/agents/relayer/src/msg/metadata/base_builder/validator_announced_storages.rs

```diff
@@ -0,0 +1,73 @@
+use eyre::Context;
+
+use hyperlane_base::cache::FunctionCallCache;
+use hyperlane_core::{ValidatorAnnounce, H256};
+
+const DOMAIN_NAME: &str = "";
+const METHOD_NAME: &str = "get_announced_storage_locations";
+
+/// Helper function to fetch storage locations for validators.
+pub async fn fetch_storage_locations_helper(
+    validators: &[H256],
+    cache: &impl FunctionCallCache,
+    validator_announce: &dyn ValidatorAnnounce,
+) -> eyre::Result<Vec<Vec<String>>> {
+    const CTX: &str = "When fetching storage locations";
+
+    let mut storage_locations = Vec::new();
+    let mut missing_validators = Vec::new();
+
+    for (index, validator) in validators.iter().enumerate() {
+        let key = generate_cache_key(validator);
+
+        // Attempt to retrieve from cache
+        if let Some(cached) = cache
+            .get_cached_call_result::<Vec<String>>(DOMAIN_NAME, METHOD_NAME, &key)
+            .await?
+        {
+            storage_locations.push(cached);
+        } else {
+            missing_validators.push((index, *validator));
+            storage_locations.push(Vec::new()); // Placeholder for missing validator
+        }
+    }
+
+    if missing_validators.is_empty() {
+        // Cache contains storage locations for all validators
+        return Ok(storage_locations);
+    }
+
+    // Fetch from validator_announce for missing validators
+    let fetched_locations = validator_announce
```

Comment:
what happens for validators that have multiple announced locations e.g. started with one location but then migrated their infra and reannounced elsewhere

Reply:
I think it'll essentially be the same behavior as before, but with some period where we have a stale value. The default expiry is 2 mins which I think is acceptable